doc/tw.in.man
.\" Originally generated by cmd.
.TH TW "1" "Monday July 2014" "tw 2014.06.07" "User Commands"
.SH NAME
tw \- Human Language Translator
.SH SYNOPSIS
.B tw [OPTIONS] [DICTIONARY] [TERM]
.SH EXAMPLE
.B tw en-es hello
.SH DESCRIPTION
translate word is a command that translates words into different languages. tw uses internal dictionaries, and contacts online to the Google Translation and the FreeTranslation engines.
.PP
Mandatory arguments for long options are mandatory for short options too.
.TP
\fB\-l\fR, \fB\-\-list\fR
list available dictionaries
.TP
\fB\-s\fR, \fB\-\-shortcuts\fR
list available dictionaries shortcuts
.TP
\fB\-x\fR, \fB\-\-exact\fR
perform an exact match
.TP
\fB\-y\fR, \fB\-\-synonyms\fR
perform a synonym triangulation
.TP
\fB\-p\fR, \fB\-\-spelling\fR
check the spelling grammar
.TP
\fB\-k\fR, \fB\-\-speak\fR
speak the text
.TP
\fB\-d\fR, \fB\-\-disable-logging\fR
disable logging of failure terms
.TP
\fB\-v\fR, \fB\-\-version\fR
show the program version
.TP
\fB\-v\fR, \fB\-\-variable\fR=\fIVARIABLE\fR
show a program variable
.TP
\fB\-h\fR, \fB\-\-help\fR
show the program help
.TP
\fB\-m\fR, \fB\-\-man\fR
show the program manual
.TP
With no term, or when term is -, read standard input. Create ~/.tw directory to enable caching and logging (mkdir ~/.tw). For straight synonyms, check mythes command.
.TP
The available --variable VARIABLES are: package, name, version, description, explanation, license, homepage, author, email, social, blog, usage, options, examples, extrahelp, extranotes, month, year, day and timestamp.
.TP
Execute 'bash -c "man <(tw --man)"' to see the runtime manpage.
.SH NOTES
Machines Content, Respect terms of online services. For more information, check documentation.
.SH HOMEPAGE
Main project page at \fBhttp://code.google.com/p/codemiscs\fR.
.SH AUTHOR
Programmed by \fBJuan Manuel Borges Caño\fR.
.SH EMAIL
Thank, Contact, Follow through \fBjuanmabcmail@gmail.com\fR.
.SH SOCIAL
Interact with \fBhttp://plus.google.com/+JuanManuelBorgesCaño\fR
.SH SUPPORT
Support through Paypal, Patreon, Flattr, Western Union, BitCoin, ...
.SH COPYRIGHT
Copyright \(co \fB2014 Juan Manuel Borges Caño\fR.
.PP
This software is licensed and under the terms of the GPLv3+ License.
There is NO WARRANTY, to the extent permitted by law.


NEWS
2014-07-03 Juan Manuel Borges Caño <juanmabcblog.blogspot.com, juanmabcmail@gmail.com> 
	* New Commands: tonize, status.
	* Improved Commands: colorize, tw.
	* Improved Website: better looking design with more easy access to stuff that matters, and more content like NEWS & ChangeLog.
	* Better Design: colorize understands pattern and line matches, removes reverse duplication, tw adapts way faster to online engines changes. Website makes core usage of templates & is better metatagged.
	* Better Documentation: continues improving effortlessly through documentation automatic syndication, automatic tonize & status man pages.
	* Better Distribution: more automated generation of documentation and wiki.
	* Better Motivation: proved benefits sweetens development, nicer web with a lot of content exploit.

2014-03-16 Juan Manuel Borges Caño <juanmabcblog.blogspot.com, juanmabcmail@gmail.com> 
	* Proved Benefits: previous features consolidate actual ones.
	* Better Motivation: proved benefits sweetens development.
	* Improved Commands: cmd (core), typewrite, colorize.
	* Improved Documentation: more rich content with media.
	* Improved Distribution: more automated generation of binary packages.
	* Better Design: typewrite can scroll, and pass through non printable characters, colorize understands more keywords.
	* Better Website: thanks to better documentation and better distribution.

2014-02-10 Juan Manuel Borges Caño <juanmabcblog.blogspot.com, juanmabcmail@gmail.com> 
	* Improved Commands: cmd (core), tw, show.
	* Better Website: Continues improving effortlessly through documentation automatic syndication and smart meta distribution. Demonstrations of every command.
	* Better Design: The cmd improvements benefit instantly all commands coding, look and usage experiences. Show plugin design is more modulable and easily expandable.
	* Better Documentation: Continues improving effortlessly through documentation automatic syndication. Code parchments got everywhere where matters.
	* Better Distribution: Introducing smart meta distribution technology (currently targeted to open build service, build.opensuse.org instance, supporting fedora, ubuntu, suse, debian, centos, arch and autotools).
	* Proved Benefits: previous releases features consolidate into real value.
	* Better Motivation: copyrighted for 2014.

2013-12-30 Juan Manuel Borges Caño <juanmabcblog.blogspot.com, juanmabcmail@gmail.com> 
	* New Commands, Improved Commands, Better Design, Better Documentation, Better Motivation.

2013-12-19 codemiscs <code.google.com/p/codemiscs>
	* New Commands, Better Documentation, Better Website, Better Distribution.

2013-11-28 codemiscs <code.google.com/p/codemiscs>
	* Independent Projects Merger.


cmds/tw/tw.in.sh
#! /usr/bin/env bash

#  _________________________________________________________________________
# /\                                                                        \
# \_|       ___        __                            _   _                  |
#   |      |_ _|_ __  / _| ___  _ __ _ __ ___   __ _| |_(_) ___  _ __       |
#   |       | || '_ \| |_ / _ \| '__| '_ ` _ \ / _` | __| |/ _ \| '_ \      |
#   |       | || | | |  _| (_) | |  | | | | | | (_| | |_| | (_) | | | |     |
#   |      |___|_| |_|_|  \___/|_|  |_| |_| |_|\__,_|\__|_|\___/|_| |_|     |
#   |                                                                       |
#   |             tw (Translate Word): Human Language Translator            |
#   |           Copyright (C) 2007 - 2014 Juan Manuel Borges Caño           |
#   |                  The need for an smart, fast and rich                 |
#   |               translation answer inspired this command.               |
#   |                    _     _                                            |
#   |                   | |   (_) ___ ___ _ __  ___  ___                    |
#   |                   | |   | |/ __/ _ \ '_ \/ __|/ _ \                   |
#   |                   | |___| | (_|  __/ | | \__ \  __/                   |
#   |                   |_____|_|\___\___|_| |_|___/\___|                   |
#   |                                                                       |
#   |  This program is free software: you can redistribute it and/or modify |
#   |  it under the terms of the GNU General Public License as published by |
#   |   the Free Software Foundation, either version 3 of the License, or   |
#   |                  (at your option) any later version.                  |
#   |                                                                       |
#   |    This program is distributed in the hope that it will be useful,    |
#   |     but WITHOUT ANY WARRANTY; without even the implied warranty of    |
#   |     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the     |
#   |              GNU General Public License for more details.             |
#   |                                                                       |
#   |   You should have received a copy of the GNU General Public License   |
#   | along with this program.  If not, see <http://www.gnu.org/licenses/>. |
#   |   ____________________________________________________________________|_
#    \_/______________________________________________________________________/

shopt -s extglob

# The --list option
function tw_option_list
{
	for tw_plugin in "$cmd_datadir/engines"/{tw_share,tw_cache,tw_local,!(tw_share|tw_cache|tw_local)}.sh
	do
		tw_twp="$tw_plugin"
		tw_twp="${tw_twp##*_}"
		tw_twp="${tw_twp%.*}"
		"tw_${tw_twp}_list" | sed -e "s/^/$("tw_${tw_twp}_name")./g"
	done
	exit 0
}

# The --shortcuts option
function tw_option_shortcuts
{
	for tw_plugin in "$cmd_datadir/engines"/{tw_share,tw_cache,tw_local,!(tw_share|tw_cache|tw_local)}.sh
	do
		tw_twp="$tw_plugin"
		tw_twp="${tw_twp##*_}"
		tw_twp="${tw_twp%.*}"
		printf "%s\n" "$("tw_${tw_twp}_name") = $("tw_${tw_twp}_shortcut")"
	done
	exit 0
}

# The --exact option
function tw_option_exact
{
	tw_exact="on"
}

# The --synonyms option
function tw_option_synonyms
{
	tw_synonyms="on"
}

# The --spelling option
function tw_option_spelling
{
	tw_spelling="on"
}

# The --speak option
function tw_option_speak
{
	tw_speak="on"
}

# The --disable-logging option
function tw_option_disable_logging
{
	tw_logging="off"
}

# The plugin loader
function tw_load_plugins
{
	for tw_plugin in "$cmd_datadir/engines"/{tw_share,tw_cache,tw_local,!(tw_share|tw_cache|tw_local)}.sh
	do
		source "$tw_plugin"
	done
}

# The spelling helper
function tw_aspell
{
	if cmd_switch "$tw_spelling"
	then
		printf "%s\n" "$tw_input" | aspell -d "${tw_dict%%-*}" -a 2> /dev/null | sed -ne '/^&/s/^& \([^ ]\+\) [0-9]\+ [0-9]\+: \(.*\)/possible misspelling of \`\1'\'', did you mean: \2?/p' | while read -r tw_msg
		do
			cmd_warning "$tw_msg"
		done
	fi
}

# The speak helper
function tw_espeak
{
	if cmd_switch "$tw_speak"
	then
		espeak -v "${tw_dict%%-*}" "$tw_input" 2> /dev/null
		[[ -n "$tw_output" ]] && espeak -v "${tw_dict##*-}" "$tw_output" 2> /dev/null
	fi
}

# The translation plugin helper: dict 
function tw
{
	if [[ -n "$tw_output" ]] || [[ -n "$tw_outputextra" ]]
	then
		tw_twd="$cmd_datadir/dicts/$tw_dict.twd"
		tw_twdc="$HOME/.tw/$tw_dict.twdc"
		if [[ -n "$tw_output" ]]
		then
			printf "%s\n" "$tw_output"
			if [[ -d "$HOME/.tw" ]] && (( "$(printf "%s\n" "$tw_input" | wc -l)" == "1" )) && (( "$(printf "%s\n" "$tw_input" | wc -w)" <= "5" ))
			then
				printf "%s\n" "$tw_output" | while read -r tw_term
				do
					if (( "$(printf "%s\n" "$tw_term" | wc -w)" <= "5" )) && ! grep -qsxi -- "$tw_input : $tw_term" "$tw_twd" "$tw_twdc"
					then
						printf "%s\n" "$tw_input : $tw_term" >> "$tw_twdc"
					fi
				done
			fi
		fi
		if [[ -n "$tw_outputextra" ]]
		then
			printf "%s\n" "$tw_outputextra"
			if cmd_switch "$tw_exactextra" && [[ -d "$HOME/.tw" ]] && (( "$(printf "%s\n" "$tw_input" | wc -l)" == "1" )) && (( "$(printf "%s\n" "$tw_input" | wc -w)" <= "5" ))
			then
				printf "%s\n" "$tw_outputextra" | while read -r tw_termextra
				do
					if (( "$(printf "%s\n" "$tw_termextra" | wc -w)" <= "5" )) && ! grep -qsxi -- "$tw_input : $tw_termextra" "$tw_twd" "$tw_twdc"
					then
						printf "%s\n" "$tw_input : $tw_termextra" >> "$tw_twdc"
					fi
				done
			fi
		fi
		tw_espeak &
	else
		if cmd_switch "$tw_logging" && [[ -d "$HOME/.tw" ]] && (( "$(printf "%s\n" "$tw_input" | wc -l)" == "1" )) && (( "$(printf "%s\n" "$tw_input" | wc -w)" <= "5" ))
		then
			tw_twdf="$HOME/.tw/$tw_dict.twdf"
			printf "%s\n" "$tw_input" >> "$tw_twdf"
			cmd_error "term not found (maybe not connected) logged $tw_twdf"
		else
			cmd_error "term not found (maybe not connected)"
		fi
	fi
	exit 0
}

# ... and tw, the program itself # The cmd init function
function tw_init
{
	shopt -s extglob
	shopt -s nullglob

	tw_load_plugins
	source "$cmd_datadir/tw_mythes.sh" 

	tw_logging="on"
	tw_exact="off"
	tw_synonyms="off"
	tw_speak="off"
	tw_spelling="off"
}

# The cmd main function
function tw_main
{
	# This checks for correct input
	if [[ -z "$1" ]]
	then
		cmd_error "unknown action"
	fi

	# This identifies the engine
	tw_engine="${1%.*}"
	if [[ "$tw_engine" = "$1" ]]
	then
		tw_engine="local"
	fi

	# This identifies the dict
	tw_dict="${1##*.}"

	# This prepares input
	if [[ -z "$2" ]] || [[ "$2" = "-" ]]
	then
		tw_input="$(cat -)"
	else
		tw_input="$2"
	fi

	# This prepares output
	tw_output=""
	tw_outputextra=""
	tw_exactextra="off"

	# This performs the plugin translation
	tw_aspell
	for tw_plugin in "$cmd_datadir/engines"/{tw_share,tw_cache,tw_local,!(tw_share|tw_cache|tw_local)}.sh
	do
		tw_twp="$tw_plugin"
		tw_twp="${tw_twp##*_}"
		tw_twp="${tw_twp%.*}"
		if [[ "$tw_engine" = "$("tw_${tw_twp}_name")" ]] || [[ "$tw_engine" = "$("tw_${tw_twp}_shortcut")" ]]
		then
			"tw_${tw_twp}"
			tw
		fi
	done
	if cmd_switch "$tw_logging" && [[ -d "$HOME/.tw" ]] && (( "$(printf "%s\n" "$tw_input" | wc -l)" == "1" )) && (( "$(printf "%s\n" "$tw_input" | wc -w)" <= "5" ))
	then
		tw_twdf="$HOME/.tw/$tw_dict.twdf"
		printf "%s\n" "$tw_input" >> "$tw_twdf"
		cmd_error "term not found (engine not found) logged $tw_twdf"
	else
		cmd_error "term not found (engine not found)"
	fi
}

# The cmd fields
cmd_package="[@]pkg[@]"
cmd="tw"
cmd_name="translate word"
cmd_version="[@]pkgversion[@]"
cmd_description="Human Language Translator"
cmd_explanation="translate word is a command that translates words into different languages. tw uses internal dictionaries, and contacts online to the Google Translation and the FreeTranslation engines."
cmd_license="[@]pkglicense[@]"
cmd_homepage="[@]pkghomepage[@]"
cmd_author="[@]pkgauthor[@]"
cmd_email="[@]pkgemail[@]"
cmd_social="[@]pkgsocial[@]"
cmd_blog="[@]pkgblog[@]"
cmd_usage="$cmd [OPTIONS] [DICTIONARY] [TERM]"
cmd_options=("/l/list/list available dictionaries/tw_option_list/" "/s/shortcuts/list available dictionaries shortcuts/tw_option_shortcuts/" "/x/exact/perform an exact match/tw_option_exact/")
if [[ "[@]pkghunspell[@]" = "yes" && "[@]pkgmythes[@]" = "yes" ]] 
then
	cmd_options=("${cmd_options[@]}" "/y/synonyms/perform a synonym triangulation/tw_option_synonyms/")
fi
if [[ "[@]pkgaspell[@]" = "yes" ]]
then
	cmd_options=("${cmd_options[@]}" "/p/spelling/check the spelling grammar/tw_option_spelling/")
fi
if [[ "[@]pkgespeak[@]" = "yes" ]]
then
	cmd_options=("${cmd_options[@]}" "/k/speak/speak the text/tw_option_speak/")
fi
cmd_options=("${cmd_options[@]}" "/d/disable-logging/disable logging of failure terms/tw_option_disable_logging/")
cmd_examples=("$cmd en-es hello")
cmd_extrahelp="With no term, or when term is -, read standard input. Create ~/.$cmd directory to enable caching and logging (mkdir ~/.$cmd)."
if [[ "[@]pkgmythes[@]" = "yes" ]]
then
	cmd_extrahelp+=" For straight synonyms, check mythes command."
fi
cmd_extranotes="Machines Content, Respect terms of online services. For more information, check documentation."
cmd_init="tw_init"
cmd_main="tw_main"

cmd_datadir="[@]pkgdatadir[@]/$cmd"

# The cmd environment
source "[@]pkgdatadir[@]/cmd.sh"
